### PR TITLE
Feature/add cmake

### PR DESCRIPTION
**Pull request by wenerme**
#1271 

Build test with this 

```Dockerfile
# Alpine with China mirror
FROM wener/base
MAINTAINER wener <wenermail@gmail.com>

# Better for cache and dev
RUN apk add --no-cache --virtual .build-deps \
    alpine-sdk cmake \
    linux-headers libev-dev libsodium-dev mbedtls-dev pcre-dev udns-dev

RUN  \
# Test this feature
    git clone --depth 1 -b feature/add-cmake https://github.com/wenerme/shadowsocks-libev /src && \
# When this feature released
#   git clone --depth 1 https://github.com/shadowsocks/shadowsocks-libev /src && \
    cd /src && git submodule update --init --recursive

RUN \
    cd /src && cmake . && make && make install
```

### CODE REVIEW

**Comment by wenerme** [comment]
There is no static udns in alpine, so in alpine we can only build share linked executable

```Dockerfile
# Alpine with China mirror
FROM wener/base
MAINTAINER wener <wenermail@gmail.com>

# Better for cache and dev
RUN apk add --no-cache --virtual .build-deps \
    alpine-sdk cmake \
    linux-headers libev-dev libsodium-dev mbedtls-static mbedtls-dev pcre-dev udns-dev

# Use echo here to invalidate cache
RUN  echo 1 &&\
# Test this feature
    git clone --depth 1 -b feature/add-cmake https://github.com/wenerme/shadowsocks-libev /src && \
# When this feature released
#   git clone --depth 1 https://github.com/shadowsocks/shadowsocks-libev /src && \
    cd /src && git submodule update --init --recursive

# Due to there is no static udns
# We do have mbedtls-static
RUN \
    cd /src && cmake -DBUILD_STATIC=OFF . && make && make install

# Cleanup
RUN \
    runDeps="$( \
        scanelf --needed --nobanner /src/bin/ss-* \
         | awk '{ gsub(/,/, "\nso:", $2); print "so:" $2 }' \
         | xargs -r apk info --installed \
         | sort -u \
    )" && \
    echo "Install runtime deps $runDeps" && \
    apk add --no-cache --virtual .run-deps $runDeps && \
    apk del .build-deps && \
    rm -rf /tmp/* && rm -rf /src
```

**Comment by registe** [comment]
good

**Comment by madeye** [comment]
Cool! There are still two missing parts here:

1. `--enable-system-shared-lib` options for linking to shared libraries of `libcork` and `libcorkipset`. It should not be that necessary for our current CMake files, as debian distribution will continue to use autoconf/automake for packaging.

2. Rules to build all the `asciidoc` in `doc` folder. It's necessary but seems not a trivial task for CMake. So, it'd be a great help if you can find a way to implement it.

File: src/config.h.in [hunk 1]
```diff
@@ -0,0 +1,306 @@
+/* config.h.in.  Generated from configure.ac by autoheader.  */
```
Comment (madeye) [inline src/config.h.in]: We'd better not include this config.h.in generated by autoconf here. If possible, please generate config.h in native CMake way.

Comment (wenerme) [inline src/config.h.in]: This is not generated by autoconf, I just copy and paste config.h.in to here, and change it to native CMake configuration file.

Comment (madeye) [inline src/config.h.in]: OK, to avoid any possible conflict between autoconf and cmake, could you move this file to the other folder, e.g. `cmake/`?

**Comment by wenerme** [comment]
1. Link shared  `libcork` and `libipset` is possible, just need to build shared library or use exists shared lib.
2. I'm not very good at cmake, I'm a javaer, but I think I can handle this, just have to find some examples.

**Comment by rogers0** [comment]
> as debian distribution will continue to use autoconf/automake for packaging.

this could be changed if cmake script is better.
actually libcork and libcorkipset are both built by cmake, in debian.

**Comment by debiansid** [comment]
Can I get PKGBUILD to make apk package for alpine Linux?

**Comment by wenerme** [comment]
> Can I get PKGBUILD to make apk package for alpine Linux?

Good idea.

**Comment by debiansid** [comment]
Before 3.0, I just copy from archlinux's PKGBUILD to APKBUILD, not sure about now.

**Comment by wenerme** [comment]
I can compile to static and dynamic now, I will figure out how to add deb/apk build, but I don't have too much time this week.

**Comment by rogers0** [comment]
@wenerme please avoid using status libraries, because it's hard to support security fixes
it's better to reply on pkg maintainer's version, by using dynamic libraries.

**Comment by vfreex** [comment]
It will be great to migrate to CMake. Looking forward to that...

**Comment by madeye** [comment]
Merged for now.

TODOs: 

1. Build documents.
2. Allow `--enable-system-shared-lib`

**Comment by qiuyuzhou** [comment]
Compiling to static is useful in some situation.

I could embed the static version ss-local into the ShadowsocksX-NG. It may be easer then current approach.

Now, I use the brew bottle version binaries with 5 dylib dependences. I can't just brew install then cp the binaries, because my mac os version is 10.12 then brew give me binaries with deployment target is 10.12. I need binaries support mac os 10.11.

So upgrade a ss-local embed in shadowsocksx-ng is painful now.

I'm a former c++ engineer. I'm familiar with cmake but not autotools. I would give a try on cmake building later.

**Comment by madeye** [comment]
@qiuyuzhou Great! Thanks for the help!

**Pull request merged**